src/jit/gentree.cpp

```diff
@@ -10166,6 +10166,18 @@ GenTreePtr                  Compiler::gtFoldExprConst(GenTreePtr tree)
                             goto LNG_OVF;
                         }
                         if ((ltemp/lval2) != lval1) goto LNG_OVF;
+#ifdef UNIX_AMD64_ABI
+                        // There is a clang 3.5 optimizer bug that in case of lval1 and lval2 equal
+                        // of 0x80000000 (MIN_LONG) the above expression (ltemp / lval2) results
+                        // in 0x80000000 (MIN_LONG. This causes an overflowing mul expression 
+                        // to be repalced with a 0, instead of throwing an overflow exception.
```

Comment:
Ah, I can see that the issue happens in release builds only. I was curious and googled a bit and by a chance found that clang supports intrinsics to perform arithmetic operations and return bool indicating if there was an overflow. For this specific case, you can use
```C
bool __builtin_smull_overflow (long x, long y, long *prod);
```
You can see all the other initrinsics for checking overflow here:
http://clang.llvm.org/docs/LanguageExtensions.html

It seems it would make sense to use them at all similar places in the Compiler::gtFoldExprConst, that means also for ADD / SUB.
It seems it could be done in a clean way by replacing the operations and check by functions that would simulate the intrinsics on windows and just call them on Unix.